test/unit/org/apache/cassandra/cql3/DurationTest.java

```diff
@@ -174,7 +175,211 @@ private long toMillis(String timeAsString)
+
+        // Test floor with a timestamp exactly equals to the start time + (1 x duration)
+        result = floorTimestamp("2016-09-27T18:12:00", "2h", "2016-09-27T16:12:00");
+        assertEquals(toMillis("2016-09-27T18:12:00"), result);
+
+        // Test floor with a timestamp in the first bucket
+        result = floorTimestamp("2016-09-27T16:13:00", "2h", "2016-09-27T16:12:00");
+        assertEquals(toMillis("2016-09-27T16:12:00"), result);
+
+        // Test floor with a timestamp in another bucket
+        result = floorTimestamp("2016-09-27T16:12:00", "2h", "2016-09-27T00:00:00");
+        assertEquals(toMillis("2016-09-27T16:00:00"), result);
+
+
+        //~ Test with a duration in days //////////////////////////////////////////////
+
+        // Test floor with a start time at the beginning of the month
+        result = floorTimestamp("2016-09-27T16:12:00", "2d", "2016-09-01T00:00:00");
+        assertEquals(toMillis("2016-09-27T00:00:00"), result);
+
+        // Test floor with a start time in the previous month
+        result = floorTimestamp("2016-09-27T16:12:00", "2d", "2016-08-01T00:00:00");
+        assertEquals(toMillis("2016-09-26T00:00:00"), result);
+
+
+        //~ Test with a duration in days and hours ////////////////////////////////////
+
+        result = floorTimestamp("2016-09-27T16:12:00", "2d12h", "2016-09-01T00:00:00");
+        assertEquals(toMillis("2016-09-26T00:00:00"), result);
+
+
+        //~ Test with a duration in months ////////////////////////////////////////////
+
+        // Test floor with a timestamp equals to the start time
+        result = floorTimestamp("2016-09-01T00:00:00", "2mo", "2016-09-01T00:00:00");
+        assertEquals(toMillis("2016-09-01T00:00:00"), result);
+
+        // Test floor with a timestamp in the first bucket
+        result = floorTimestamp("2016-09-27T16:12:00", "2mo", "2016-09-01T00:00:00");
+        assertEquals(toMillis("2016-09-01T00:00:00"), result);
+
+        // Test floor with a start time at the beginning of the year (LEAP YEAR)
+        result = floorTimestamp("2016-09-27T16:12:00", "1mo", "2016-01-01T00:00:00");
+        assertEquals(toMillis("2016-09-01T00:00:00"), result);
+
+        // Test floor with a start time at the beginning of the previous year
+        result = floorTimestamp("2016-09-27T16:12:00", "2mo", "2015-01-01T00:00:00");
+        assertEquals(toMillis("2016-09-01T00:00:00"), result);
+
+        // Test floor with a start time in the previous year
+        result = floorTimestamp("2016-09-27T16:12:00", "2mo", "2015-02-02T00:00:00");
+        assertEquals(toMillis("2016-08-02T00:00:00"), result);
+
+
+        //~ Test with a duration in months and days ////////////////////////////////////
+
+        // Test floor with a start time at the beginning of the year (LEAP YEAR)
+        result = floorTimestamp("2016-09-27T16:12:00", "2mo2d", "2016-01-01T00:00:00");
+        assertEquals(toMillis("2016-09-09T00:00:00"), result);
```

Comment:
nit: feels interested to add a test case that has more than 2 fields in duration specified. e.g.
```
        result = floorTimestamp("2016-09-27T16:12:00", "2mo2d8h", "2016-01-01T00:00:00");
        assertEquals(toMillis("2016-09-10T08:00:00"), result);
```